File: problems/983_minimum_cost_for_tickets/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        unordered_set<int> daySet(days.begin(), days.end());
        int last = days.back();
        vector<int> dp(last + 1, 0);
        for (int i = 1; i <= last; i++) {
            if (!daySet.count(i)) { dp[i] = dp[i - 1]; continue; }
            dp[i] = min({dp[i - 1] + costs[0],
                         dp[max(0, i - 7)] + costs[1],
                         dp[max(0, i - 30)] + costs[2]});
        }
        return dp[last];
    }
};

File: problems/983_minimum_cost_for_tickets/Solution.java
import java.util.*;

class Solution {
    public int mincostTickets(int[] days, int[] costs) {
        Set<Integer> daySet = new HashSet<>();
        for (int d : days) daySet.add(d);
        int last = days[days.length - 1];
        int[] dp = new int[last + 1];
        for (int i = 1; i <= last; i++) {
            if (!daySet.contains(i)) { dp[i] = dp[i - 1]; continue; }
            dp[i] = Math.min(dp[i - 1] + costs[0],
                    Math.min(dp[Math.max(0, i - 7)] + costs[1],
                             dp[Math.max(0, i - 30)] + costs[2]));
        }
        return dp[last];
    }
}
